File: mobile/ios/AuthSample.swift
import Foundation

struct TokenPair: Codable {
    let access: String
    let refresh: String
}

protocol TokenStore {
    func save(_ tokens: TokenPair)
    func load() -> TokenPair?
}

final class InMemoryTokenStore: TokenStore {
    private var tokens: TokenPair?

    func save(_ tokens: TokenPair) {
        self.tokens = tokens
    }

    func load() -> TokenPair? {
        tokens
    }
}

struct AuthClient {
    let baseURL: URL

    func login(username: String, password: String) async throws -> TokenPair {
        let url = baseURL.appendingPathComponent("/api/v1/auth/token/")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let payload = ["username": username, "password": password]
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)

        let (data, response) = try await URLSession.shared.data(for: request)
        if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode >= 400 {
            throw URLError(.badServerResponse)
        }

        return try JSONDecoder().decode(TokenPair.self, from: data)
    }

    func refresh(refreshToken: String) async throws -> TokenPair {
        let url = baseURL.appendingPathComponent("/api/v1/auth/token/refresh/")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let payload = ["refresh": refreshToken]
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)

        let (data, response) = try await URLSession.shared.data(for: request)
        if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode >= 400 {
            throw URLError(.badServerResponse)
        }

        return try JSONDecoder().decode(TokenPair.self, from: data)
    }
}

// Example usage (wire into your auth flow).
//
// let auth = AuthClient(baseURL: URL(string: "https://api.assetra.example")!)
// let tokens = try await auth.login(username: "user", password: "pass")
// tokenStore.save(tokens)
// let refreshed = try await auth.refresh(refreshToken: tokens.refresh)
// tokenStore.save(refreshed)

File: mobile/android/AuthSample.kt
package com.assetra.auth

import java.net.HttpURLConnection
import java.net.URL
import org.json.JSONObject


data class TokenPair(
    val access: String,
    val refresh: String,
)

interface TokenStore {
    fun save(tokens: TokenPair)
    fun load(): TokenPair?
}

class InMemoryTokenStore : TokenStore {
    private var tokens: TokenPair? = null

    override fun save(tokens: TokenPair) {
        this.tokens = tokens
    }

    override fun load(): TokenPair? = tokens
}

class AuthClient(private val baseUrl: String) {
    fun login(username: String, password: String): TokenPair {
        val url = URL("${baseUrl.trimEnd('/')}/api/v1/auth/token/")
        val payload = JSONObject()
        payload.put("username", username)
        payload.put("password", password)
        return postJson(url, payload)
    }

    fun refresh(refreshToken: String): TokenPair {
        val url = URL("${baseUrl.trimEnd('/')}/api/v1/auth/token/refresh/")
        val payload = JSONObject()
        payload.put("refresh", refreshToken)
        return postJson(url, payload)
    }

    private fun postJson(url: URL, payload: JSONObject): TokenPair {
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        connection.setRequestProperty("Content-Type", "application/json")
        connection.doOutput = true

        connection.outputStream.use { stream ->
            stream.write(payload.toString().toByteArray())
        }

        val responseText = connection.inputStream.bufferedReader().use { it.readText() }
        val responseJson = JSONObject(responseText)
        return TokenPair(
            access = responseJson.getString("access"),
            refresh = responseJson.getString("refresh"),
        )
    }
}

// Example usage (wire into your auth flow).
//
// val auth = AuthClient(baseUrl = "https://api.assetra.example")
// val tokens = auth.login("user", "pass")
// tokenStore.save(tokens)
// val refreshed = auth.refresh(tokens.refresh)
// tokenStore.save(refreshed)
